warp10/src/main/java/io/warp10/script/functions/ADDDURATION.java

```diff
@@ -0,0 +1,151 @@
+//   limitations under the License.
+//
+
+package io.warp10.script.functions;
+
+import io.warp10.continuum.gts.UnsafeString;
+import io.warp10.continuum.store.Constants;
+import io.warp10.script.NamedWarpScriptFunction;
+import io.warp10.script.WarpScriptException;
+import io.warp10.script.WarpScriptLib;
+import io.warp10.script.WarpScriptStack;
+import io.warp10.script.WarpScriptStackFunction;
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+import org.joda.time.MutablePeriod;
+import org.joda.time.ReadWritablePeriod;
+import org.joda.time.format.ISOPeriodFormat;
+
+import java.util.List;
+import java.util.Locale;
+
+public class ADDDURATION extends NamedWarpScriptFunction implements WarpScriptStackFunction {
+
+  final private static WarpScriptStackFunction TSELEMENTS = new TSELEMENTS(WarpScriptLib.TSELEMENTS);
+  final private static WarpScriptStackFunction FROMTSELEMENTS = new FROMTSELEMENTS(WarpScriptLib.TSELEMENTSTO);
+
+  public ADDDURATION(String name) {
+    super(name);
+  }
+
+  @Override
+  public WarpScriptStack apply(WarpScriptStack stack) throws WarpScriptException {
+
+    //
+    // Retrieve arguments
+    //
+
+    Object top = stack.pop();
+
+    if (!(top instanceof String || top instanceof Long)) {
+      throw new WarpScriptException(getName() + " expects an ISO8601 duration (a string) on top of the stack (see http://en.wikipedia.org/wiki/ISO_8601#Durations), or a number of durations (a Long).");
+    }
+
+    String duration;
+    long N = 1;
+    if (top instanceof String) {
+      duration = top.toString();
+    } else {
+      N = (Long) top;
+      top = stack.pop();
+      if (!(top instanceof String)) {
+        throw new WarpScriptException(getName() + " expects an ISO8601 duration (a string) in the second level of the stack (see http://en.wikipedia.org/wiki/ISO_8601#Durations).");
+      }
+      duration = top.toString();
+    }
+
+    String tz = null;
+    if (stack.peek() instanceof String) {
+      tz = stack.pop().toString();
+      if (!(stack.peek() instanceof Long)) {
+        throw new WarpScriptException(getName() + " operates on a tselements list, timestamp, or timestamp and timezone.");
+      }
+    } else if (!(stack.peek() instanceof List || stack.peek() instanceof Long)) {
+      throw new WarpScriptException(getName() + " operates on a tselements list, timestamp, or timestamp and timezone.");
+    }
+
+    //
+    // Handle duration
+    //
+
+    // Separate seconds from  digits below second precision
+    String[] tokens = UnsafeString.split(duration, '.');
+
+    long offset = 0;
+    if (tokens.length > 2) {
+      throw new WarpScriptException(getName() + "received an invalid ISO8601 duration.");
+    }
+
+    if (2 == tokens.length) {
+      duration = tokens[0].concat("S");
+      String tmp = tokens[1].substring(0, tokens[1].length() - 1);
+      Double d_offset = Double.valueOf("0." + tmp) * new Double(Constants.TIME_UNITS_PER_S);
```

Comment:
Actually you could create a Double constant instead of creating a new Double on each call.